File: backend/app/entities/stock.py
from __future__ import annotations

from sqlalchemy import Integer, Text
from sqlalchemy.orm import Mapped, mapped_column

from backend.app.core.database import Base


class Stock(Base):
    __tablename__ = "stocks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    stock_code: Mapped[str] = mapped_column(Text, nullable=False, unique=True)
    stock_name: Mapped[str] = mapped_column(Text, nullable=False)
    market: Mapped[str | None] = mapped_column(Text, nullable=True)
    sector: Mapped[str | None] = mapped_column(Text, nullable=True)
    industry: Mapped[str | None] = mapped_column(Text, nullable=True)
    isin_code: Mapped[str | None] = mapped_column(Text, nullable=True)
    corp_name: Mapped[str | None] = mapped_column(Text, nullable=True)
    corp_reg_no: Mapped[str | None] = mapped_column(Text, nullable=True)
    last_synced_at: Mapped[str | None] = mapped_column(Text, nullable=True)
    source: Mapped[str | None] = mapped_column(Text, nullable=True)
    security_type: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[str] = mapped_column(Text, nullable=False)
    updated_at: Mapped[str] = mapped_column(Text, nullable=False)

File: frontend/src/types/stock.ts
export type Stock = {
  id: number;
  stock_code: string;
  stock_name: string;
  market: string | null;
  sector: string | null;
  industry: string | null;
  isin_code: string | null;
  corp_name: string | null;
  corp_reg_no: string | null;
  last_synced_at: string | null;
  source: string | null;
  security_type: string | null;
  is_active: number;
  created_at: string;
  updated_at: string;
};

export type StockListParams = {
  keyword?: string;
  is_active?: number;
  market?: string;
  security_type?: string;
  limit?: number;
  offset?: number;
};

export type StockCreateInput = {
  stock_code: string;
  stock_name: string;
  market?: string;
  sector?: string;
  industry?: string;
  security_type?: string;
};

export type StockUpdateInput = {
  stock_code?: string;
  stock_name?: string;
  market?: string;
  sector?: string;
  industry?: string;
  security_type?: string;
  is_active?: number;
};

export type StockCodeNormalizeItem = {
  stock_id: number;
  stock_name: string;
  old_code: string;
  new_code: string;
  status: string;
};

export type StockCodeNormalizeResponse = {
  dry_run: boolean;
  target_count: number;
  updated_count: number;
  duplicate_conflict_count: number;
  items: StockCodeNormalizeItem[];
};
